python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,652 @@
+
+        n_rows = 100
+        more_data = cp.random.random_sample((n_rows, n_features), dtype=cp.float32)
+        indices = index.size + cp.arange(n_rows, dtype=cp.uint64)
+        index = ivf_pq.extend(index, more_data, indices)
+
+        # Search using the built index
+        queries = cp.random.random_sample((n_queries, n_features), dtype=cp.float32)
+        k = 10
+        neighbors = cp.empty((n_queries, k), dtype=cp.uint64)
+        distances = cp.empty((n_queries, k), dtype=cp.float32)
+        ivf_pq.search(ivf_pq.SearchParams(), index, queries, k, neighbors, distances, handle=handle)
+
+        # pylibraft functions are often asynchronous so the
+        # handle needs to be explicitly synchronized
+        handle.sync()
+    
+    """
+    if not index.trained:
+        raise ValueError("Index need to be built before calling extend.")
+
+    if handle is None:
+        handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+
+    vecs_cai = new_vectors.__cuda_array_interface__
+    vecs_dt = np.dtype(vecs_cai["typestr"])
+    cdef uint64_t n_rows = vecs_cai["shape"][0]
+    cdef uint32_t dim = vecs_cai["shape"][1]
+
+    _check_input_array(vecs_cai, [np.dtype('float32'), np.dtype('byte'), np.dtype('ubyte')], 
+                       exp_cols=index.dim)
+
+    idx_cai = new_indices.__cuda_array_interface__
+    _check_input_array(idx_cai, [np.dtype('uint64')], exp_rows=n_rows)
+    if len(idx_cai["shape"])!=1:
+        raise ValueError("Indices array is expected to be 1D")
+
+
+    cdef uintptr_t vecs_ptr = vecs_cai["data"][0]
+    cdef uintptr_t idx_ptr = idx_cai["data"][0]
+
+    if vecs_dt == np.float32:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <float*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)
+    elif vecs_dt == np.int8:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <int8_t*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)
+    elif vecs_dt == np.uint8:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <uint8_t*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)       
+    else:
+        raise TypeError("query dtype %s not supported" % vecs_dt)
+
+    return index 
+
+
+cdef class SearchParams:
+    cdef c_ivf_pq.search_params params
+
+    def __init__(self, *, n_probes=20, 
+                 lut_dtype=CUDA_R_32F, 
+                 internal_distance_dtype=CUDA_R_32F):
+        """
+        IVF-PQ search parameters
+
+        Parameters
+        ----------
+        n_probes: int, default = 1024
+            The number of course clusters to select for the fine search.
+        lut_dtype: default = ivf_pq.CUDA_R_32F (float)
```

Comment:
We should accept numpy dtypes here and convert them into the corresponding cuda data types (applies to all parts that a user sets explicitly)